spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/Mqttv5PahoMessageDrivenChannelAdapter.java

```diff
@@ -341,6 +380,46 @@ public void authPacketArrived(int reasonCode, MqttProperties properties) {
 
 	}
 
+	private void subscribeToAll() {
+		ApplicationEventPublisher applicationEventPublisher = getApplicationEventPublisher();
+		String[] topics = getTopic();
+		this.topicLock.lock();
+		try {
+			if (topics.length == 0) {
+				return;
+			}
+
+			int[] requestedQos = getQos();
+			if (this.mqttClient != null) {
```

Comment:
Cannot we have something like `this.mqttClient = getClientManager().getClient()` and have the rest of logic really relying on the `this.mqttClient` without such an overhaul with extra `if..else` ?

Reply:
Oh, I'd like to get rid of all these if/else. That's why initially I was going to use different implementations.
The issue here is that 
1. `clientManager` has its own lifecycle, it should not do connect/disconnect in adapters/handlers
2. `clientManager` approach uses its own subscription logic (different method with listeners)

Theoretically, we could make old client use `MessageListener`-based subscription as well, but who knows how this will work comparing to the current behavior. If you think we can accept this risk, I think I could make it more similar. Anyway, we still have difference #1 and the number of `if`s will be just minus one...

Reply:
Re. `#2` - yes, that's my point. We move the logic from the common callback to this listener and the end logic is going to be the same.
The code in the client looks like this:
```
mqttCallback.deliveryComplete(token);
...
	MqttActionListener asyncCB = token.getActionCallback();
			if (asyncCB != null) {
				if (token.getException() == null) {
					// @TRACE 716=call onSuccess key={0}
					log.fine(CLASS_NAME, methodName, "716", new Object[] { token.internalTok.getKey() });
					asyncCB.onSuccess(token);
				} else {
					// @TRACE 717=call onFailure key {0}
					log.fine(CLASS_NAME, methodName, "716", new Object[] { token.internalTok.getKey() });
					asyncCB.onFailure(token, token.getException());
				}
			}
```

So, it calls the common callback and then action listener if present.

Right. The first item has a flaw that we need to check for the external client manager before performing a lifecycle action on the client property.
But that's it: the rest of the logic is going to be around `this.client` and should be exactly the same as it is this right now with your `getClientManager().getClient()`

Reply:
I will try to make the difference only on lifecycle methods (start/stop) and verify via existing auto tests. Thanks!